manager/orchestrator/restart.go

```diff
@@ -57,11 +57,49 @@ func NewRestartSupervisor(store *store.MemoryStore) *RestartSupervisor {
 	}
 }
 
+func (r *RestartSupervisor) waitRestart(ctx context.Context, oldDelay delayedStart, cluster *api.Cluster, taskID string) {
+	// Wait for the last restart delay to elapse.
+	select {
+	case <-oldDelay.doneCh:
+	case <-ctx.Done():
+		return
+	}
+
+	// Start the next restart
+	err := r.store.Update(func(tx store.Tx) error {
+		t := store.GetTask(tx, taskID)
+		if t == nil {
+			return nil
+		}
+		service := store.GetService(tx, t.ServiceID)
+		if service == nil {
+			return nil
+		}
+		return r.Restart(ctx, tx, cluster, service, *t)
+	})
+
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to restart task after waiting for previous restart")
+	}
+}
+
 // Restart initiates a new task to replace t if appropriate under the service's
 // restart policy.
 func (r *RestartSupervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Cluster, service *api.Service, t api.Task) error {
 	// TODO(aluzzardi): This function should not depend on `service`.
 
+	// Is the old task still in the process of restarting? If so, wait for
+	// its restart delay to elapse, to avoid tight restart loops (for
+	// example, when the image doesn't exist).
+	r.mu.Lock()
+	oldDelay, ok := r.delays[t.ID]
+	if ok {
+		go r.waitRestart(ctx, oldDelay, cluster, t.ID)
```

Comment:
I thought this would not be an issue because because `Restart` only gets when a task transitions to an observed state `> RUNNING` (i.e., it has failed or completed). But I suppose it's possible that a task could first go to COMPLETED and then SHUTDOWN, for example, and that would trigger two goroutines. So I guess I'll add something here to limit it to one goroutine.